server/aesdsocket.c

```diff
@@ -0,0 +1,287 @@
+
+            dup2(dev_null_fd, STDIN_FILENO);
+            dup2(dev_null_fd, STDOUT_FILENO);
+            dup2(dev_null_fd, STDERR_FILENO);
+
+            close(STDIN_FILENO);
+            close(STDOUT_FILENO);
+            close(STDERR_FILENO);
+        }
+    }
+
+	// listen on socket
+    status = listen(socket_num, MAX_BACKLOG);
+    if (status == -1) {
+        perror("listen");
+        return -1;
+    }
+	
+	// open file for writing
+    file_fd = open(OUTPUT_FILE_PATH, O_CREAT | O_APPEND | O_RDWR, S_IRWXU);
+    if (file_fd == -1) {
+        perror("open");
+        return -1;
+    }
+
+    int num_bytes;
+    char buf[HUNDRED_BYTES];
+	memset(buf, '\0', HUNDRED_BYTES); // clear buf
+	int send_buf_size = 0;
+
+	while (1) {
+        // accept connection from client
+        client_fd = accept(socket_num, (struct sockaddr*) &client_sockaddr, &client_addrlen);
+        if (client_fd == -1) {
+            perror("accept");
+            return -1;
+        }
+
+		// log IP to syslog
+        char* ip_addr = inet_ntoa(client_sockaddr.sin_addr);
+        syslog(LOG_DEBUG, "Accepted connection from %s\n", ip_addr);
+
+		int recv_buf_pos = 0;
+		long recv_buf_size = HUNDRED_BYTES;
+		
+		char* recv_buf = malloc(recv_buf_size * sizeof(char));
+        if (recv_buf == NULL) {
+            perror("malloc failure");
+            return -1;
+        }
+
+        // mask signals
+        status = sigprocmask(SIG_BLOCK, &cur_set, &prev_set);
+        if (status == -1) {
+            printf("signal masking failed\n");
+        }
+
+        // receive data from client
+        // each loop iteration is 1 newline of input
+        while (1) {
+            // save up to 100 bytes into buf
+			num_bytes = recv(client_fd, buf, HUNDRED_BYTES, 0);
+
+            // exit loop on error
+            if (num_bytes == -1) {
+                perror("recv");
+                return -1;
+            }
+
+			// check if allocated buf size is sufficient
+			if(recv_buf_pos + num_bytes > recv_buf_size) {
+				recv_buf_size += num_bytes;
+				recv_buf = realloc(recv_buf, recv_buf_size * sizeof(char));
+			}
+			
+			// copy buf to recv_buf
+			memcpy(recv_buf + recv_buf_pos, buf, num_bytes);
+			recv_buf_pos += num_bytes;
+
+            // exit loop if there was a new line character received
+			if(strchr(buf, '\n') != NULL)
+				break;
+        }
+
+        // unmask signals
+        status = sigprocmask(SIG_UNBLOCK, &prev_set, NULL);
+        if (status == -1) {
+            printf("signal unmasking failed\n");
+        }
+		
+		// write new bytes to file
+		num_bytes = write(file_fd, recv_buf, recv_buf_pos);
+        printf("** Writing %d bytes\n", num_bytes);
+		if (num_bytes == -1 || num_bytes != recv_buf_pos) {
+			perror("write");
+			return -1;
+		}
+		
+		// go to start of file
+		lseek(file_fd, 0, SEEK_SET);
+		
+		// allocate memory for send buffer
+		send_buf_size += recv_buf_pos;
+		char* send_buf = malloc(send_buf_size * sizeof(char));
```

Comment:
@bjornhnelson what assumption does this make about the size of RAM heap realtive to the total size of all packets sent?